### PR TITLE
Cross-platform "unixtools"

### PR DESCRIPTION
**Pull request by matthewbauer**
## Motivation for this change
This adds compatibility for procps, nettools, and utillinux to non-Linux environments. See #25188 for more info. Still needs better documentation.

### CODE REVIEW

**Comment by Mic92** [comment]
How do we teach people to use `unixtools` instead of `utillinux` people already know?

**Comment by 7c6f434c** [comment]
Maybe it is better to have a `unixtools` derivation that includes _only_ the tools provided on all supported platforms? Debugging macOS builds without access to macOS is annoying, so a package explicitly intended to be a compatibility layer should probably aim to behave in the same way everywhere.

**Comment by matthewbauer** [comment]
> How do we teach people to use unixtools instead of utillinux people already know? Would it be simpler to just call it utilinux?

I guess that's possible. It's a little bit inaccurate because of how many things are unavailable in BSDs:

https://en.wikipedia.org/wiki/Util-linux

> Maybe it is better to have a unixtools derivation that includes only the tools provided on all supported platforms?

That seems like a good idea, and something I think we can do with buildEnv.

> An alternative to have one set of packages would be to explicit: for example having an attribute called kill.

So I thought about this and I think for kill and ps at least it makes sense. We probably still want some sort of compatibility for procps, etc.

**Comment by matthewbauer** [comment]
Ok so I've redone this PR. Interested in whether this breaks anything on Linux. So far so good! Still needs documentation though.

File: pkgs/top-level/unix-tools.nix [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    if builtins.hasAttr hostPlatform.parsed.kernel.name providers then
+      runCommand cmd {} ''
+        mkdir -p $out/bin
+
+        if ! [ -x "${providers.${hostPlatform.parsed.kernel.name}}/bin/${cmd}" ]; then
+          echo "Cannot find command ${cmd}"
+          exit 1
+        fi
+
+        ln -s ${providers.${hostPlatform.parsed.kernel.name}}/bin/${cmd} $out/bin/${cmd}
```
Comment (matthewbauer) [inline pkgs/top-level/unix-tools.nix]: @Ericson2314 Is there a better way to get this? Basically I want to get the second part of the "x86_64-darwin" tuple.

Comment (Ericson2314) [reply pkgs/top-level/unix-tools.nix]: Don't think so! haha. Maybe eventually `parsed` will contain everything and then we can remove the indirection.

**Review by Ericson2314 — APPROVED** [review]
This is a really great change! Not sure how it looked like before but really like this final iteration. Not giving anyone "extra programs for free" helps keep us honest too.

**Comment by Ericson2314** [comment]
@7c6f434c We can maybe put warnings on the originals, telling people to use these instead?

**Comment by matthewbauer** [comment]
>We can maybe put warnings on the originals, telling people to use these instead?

That's probably not going to work for certain cases. For instance ```unshare``` only exists in Linux AFAIK. I think it might make more sense to just refer to util-linux for that (I guess we could provide "unshare" here and throw an error on non-Linux).

There are definitely a lot more commands that we could add but are unavailable on Darwin for various reasons (I haven't been able to get pkill or pgrep to build correctly).

**Comment by Ericson2314** [comment]
@matthewbauer Yeah even if its missing some places (or like `unshare` unlikely to be ported) I think its better to list them one by one.

**Comment by 7c6f434c** [comment]
Re: warnings: these warnings will also happen for user environment construction, though, which is not desirable. I think that we need to create a stalled RFC (I have no illusions here) to gather enough support to just slowly convert most cases. If the RFC gets little support of gets actually closed, then we know this won't work…

(Note that sometimes it does make sense to use full util-linux for a build, too…)

**Comment by Mic92** [comment]
@7c6f434c I do not understand what you want to make a RFC about?

File: pkgs/top-level/unix-tools.nix (hunk 1 above)
Comment (Mic92) [reply pkgs/top-level/unix-tools.nix]: Could this be a normal `stdenv.derivation` with `meta.platforms` with regards to `borg` and other tooling? or is throw handled in the same way as `meta.platforms`?

**Comment by Mic92** [comment]
The new version looks much better and future proof (with regards to new platforms).

**Comment by 7c6f434c** [comment]
Deprecate usage of `utillinux` and `procps` as build inputs for packages in the Nixpkgs mainline.

**Comment by Mic92** [comment]
@7c6f434c I don't think we need an RFC for that. There are not really deprecated. This is just an abstraction that uses either `utillinux` or something else depending on the platform to have less nix boiler code and less breakages, if macOS support breaks by people adding `utillinux` to `buildInputs`.

**Comment by Ericson2314** [comment]
@Mic92 I believe you two are in agreement. @7c6f434c is saying that the *possible* next step of deprecating the old ones would require an RFC. Nobody thinks this PR needs an RFC.

File: pkgs/top-level/unix-tools.nix [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+# You should always try to use single binaries when available. For
+# instance, if your program needs to use "ps", just list it as a build
+# input, not "procps" which requires Linux.
+
+let
+
+  singleBinary = cmd: providers:
+    if builtins.hasAttr hostPlatform.parsed.kernel.name providers then
+      runCommand cmd {} ''
```
Comment (matthewbauer) [inline pkgs/top-level/unix-tools.nix]: I'd like to do a passthru var here that points to "$out/bin/${cmd}". Does anyone know if this is possible (without placeholder).

File: pkgs/top-level/all-packages.nix [hunk 3]
```diff
@@ -13096,16 +13086,14 @@ with pkgs;
     })
     else if stdenv.isLinux
     then utillinuxMinimal
-    else null;
+    else libossp_uuid;
```
Comment (matthewbauer) [inline pkgs/top-level/all-packages.nix]: I'm hoping these are completely compatible? Ideally I'd like libossp_uuid to be used everywhere but not sure if that will break something on Linux.

Comment (matthewbauer) [reply pkgs/top-level/all-packages.nix]: I'll probably just remove this and put it in another PR.

File: pkgs/top-level/unix-tools.nix (hunk 1 above)
Comment (matthewbauer) [reply pkgs/top-level/unix-tools.nix]: "throw" is allowed to be a little more descriptive which in this case seems good. I think in the end meta.platforms calls throw if system doesn't match.

**Comment by matthewbauer** [comment]
I will write a section in the manual under "Packaging in a cross-friendly manner" tonight if I find the time.

But what are opinions on this? I'd like to get it merged into staging soonish but also keep working on some of the edge cases (especially with adding new "tools"). This shouldn't break anything but just want to make sure people are okay with this PR (or at least not strongly against).

**Review by LnL7 — APPROVED** [review]

**Comment by LnL7** [comment]
I really like the idea of making the commands explicit instead of using package collections like utillinux that don't work across multiple platforms. I'm not entirely sure if it's a good idea to expose the fallback collections top-level since not everything is in there but it's probably fine is most cases.

File: pkgs/top-level/unix-tools.nix (hunk 1 above)
Comment (Ericson2314) [reply pkgs/top-level/unix-tools.nix]: Ping me and I'll fix it after that is merged :).

**Comment by matthewbauer** [comment]
Merging now, so @Ericson2314 can make his changes. Also working on some documentation but that will go in another PR. (sorry for the large merge!)

**Comment by Mic92** [comment]
Is there a pull request for documentation?

**Comment by matthewbauer** [comment]
Not yet.

**Comment by dtzWill** [comment]
FWIW looks like `hexdump` should come from `utillinux` (or `utillinuxMinimal`? not sure), it's not available in `procps` where we currently try to grab it from.

(I'd put together a PR but am currently rebuilding all the things to get to the point where I can test such.. haha)

**Comment by matthewbauer** [comment]
Ok thanks! I hope none of these tools are used in bootstrapping stdenv. My understanding is utillinuxMinimal only makes sense in those cases (although it does have hexdump, so maybe it would shrink dependencies).

**Pull request merged**